Generators/include/Generators/GeneratorPythia8.h

```diff
@@ -88,27 +166,101 @@ class GeneratorPythia8 : public Generator
   /** operator= **/
   GeneratorPythia8& operator=(const GeneratorPythia8&);
 
-  /** methods that can be overridded **/
-  void updateHeader(o2::dataformats::MCEventHeader* eventHeader) override;
+  /** @{
+   * @name Some function definitions
+   */
+  /** Select particles when pruning event */
+  typedef bool (*Select)(const Pythia8::Particle& particle);
+  /** Get relatives (mothers or daughters) of a particle */
+  typedef std::vector<int> (*GetRelatives)(const Pythia8::Particle&);
+  /** Set relatives (mothers or daughters) of a particle */
+  typedef void (*SetRelatives)(Pythia8::Particle&, int, int);
+  /** Get range of relatives (mothers or daughters) of a particle */
+  typedef std::pair<int, int> (*FirstLastRelative)(const Pythia8::Particle&);
+  /** @} */
 
-  /** internal methods **/
-  Bool_t importParticles(Pythia8::Event const& event);
+  /** @{
+   * @name Methods that can be overridded **/
+  /** Update the event header.  This propagates all sorts of
+   * information from Pythia8 to the simulation event header,
+   * including parton distribution function parameters, event weight,
+   * cross-section information, heavy-ion collision information, and
+   * so on. */
+  void updateHeader(o2::dataformats::MCEventHeader* eventHeader) override;
+  /** @} */
 
-  /** utilities **/
+  /** @{
+   * @name Internal methods **/
+  /** Import particles from Pythia onto the simulation stack */
+  Bool_t importParticles(Pythia8::Event& event);
+  /** Prune an event.  Only particles for which the function select
+   * returns true are kept in the event record.  The structure of the
+   * event is preserved, meaning that particles will point back to
+   * their ultimate (selected) mothers and, if select preserves the
+   * beam particles, the ultimate collision interaction. */
+  void pruneEvent(Pythia8::Event& event, Select select);
+  /** Investigate relatives (mothers or daughters) for particles to
+   * keep when pruning an event.  This checks the current particle,
+   * identified by index, if any of its relatives (either mothers or
+   * daughters) are to be kept.  If a relative is to be kept, then
+   * that is added to an internal list.  If a relative is _not_ to be
+   * kept, then that relatives relatives are queried (recursive call).
+   * The result of the recursive call is a list of relatives to the
+   * current particle which are to be kept.  These are then also added
+   * to the internal list.  The relatives that are found to be kept
+   * are then set to be relatives of the current particle.  Note that
+   * this member function modifies the relatives of the passed
+   * particle, and thus modifies the passed event structure.
+   * Calculations are cached. */
+  void investigateRelatives(Pythia8::Event& event,                   // Event
+                            const std::map<size_t, size_t>& old2New, // Map from old to new idx
```

Comment:
Part of the reason why the pruning became much slower might be the use of std::map<> and std::set<>. Access to these scales like O(log(N)) and I would typically advise the use of unordered_map<> for both of these cases (O(1) access).

Reply:
See longer reply below